syne_tune/optimizer/schedulers/searchers/kde_searcher.py

```diff
@@ -0,0 +1,275 @@
+    bandwidth_factor: int
+        We sample continuous hyperparameter from a truncated Normal. This factor is multiplied to the bandwidth to
+        define the standard deviation of this trunacted Normal.
+    random_fraction: float
+        Defines the fraction of configurations that are drawn uniformly at random instead of sampling from the model
+    points_to_evaluate: List[Dict] or None
+        List of configurations to be evaluated initially (in that order).
+        Each config in the list can be partially specified, or even be an
+        empty dict. For each hyperparameter not specified, the default value
+        is determined using a midpoint heuristic.
+        If None (default), this is mapped to [dict()], a single default config
+        determined by the midpoint heuristic. If [] (empty list), no initial
+        configurations are specified.
+    """
+
+    def __init__(
+            self,
+            configspace: Dict,
+            metric: str,
+            mode: str = "min",
+            num_min_data_points: int = None,
+            top_n_percent: int = 15,
+            min_bandwidth: float = 1e-3,
+            num_candidates: int = 64,
+            bandwidth_factor: int = 3,
+            random_fraction: float = .33,
+            points_to_evaluate: Optional[List[Dict]] = None,
+            **kwargs
+    ):
+        super().__init__(configspace=configspace, metric=metric, points_to_evaluate=points_to_evaluate)
+        self.mode = mode
+        self.metric_name = metric
+        self.num_evaluations = 0
+        self.min_bandwidth = min_bandwidth
+        self.random_fraction = random_fraction
+        self.num_candidates = num_candidates
+        self.bandwidth_factor = bandwidth_factor
+        self.points_to_evaluate = points_to_evaluate
+        self.top_n_percent = top_n_percent
+        self.X = []
+        self.y = []
+        self.categorical_maps = {
+            k: {cat: i for i, cat in enumerate(v.categories)}
+            for k, v in configspace.items()
+            if isinstance(v, sp.Categorical)
+        }
+        self.inv_categorical_maps = {
+            hp: dict(zip(map.values(), map.keys())) for hp, map in self.categorical_maps.items()
+        }
+
+        self.good_kde = None
+        self.bad_kde = None
+
+        self.vartypes = []
+        for hname in self.configspace:
+            hp = self.configspace[hname]
+            if isinstance(hp, sp.Categorical):
+                self.vartypes.append(('u', len(hp.categories)))
+            if isinstance(hp, sp.Integer):
+                self.vartypes.append(('o', (hp.lower, hp.upper)))
+            if isinstance(hp, sp.Float):
+                self.vartypes.append(('c', 0))
+
+        self.num_min_data_points = len(self.vartypes) if num_min_data_points is None else num_min_data_points
+        assert self.num_min_data_points >= len(self.vartypes)
+
+    @staticmethod
+    def to_feature(configspace, config, categorical_maps):
```

Comment:
It would be nicer to use HyperparameterRanges here. But I see this would be a small step away from the original code.

One thing: For int type, lower is always mapped to 0, upper to 1. Is that what you want? The default encoding of int a:b to real [0, 1] is rather mapping a to 1/(2 n), where n = b - a + 1. This makes sure that when you sample from [0, 1], then round to nearest int, this is the same as sampling from int a:b directly.

Reply:
good catch